src/config/options.rs

```diff
@@ -105,11 +105,68 @@ macro_rules! configuration_option_enum{
 }
 
 configuration_option_enum! { NewlineStyle:
+    Auto, // Auto-detect based on the raw source input
     Windows, // \r\n
     Unix, // \n
     Native, // \r\n in Windows, \n on other platforms
 }
 
+impl NewlineStyle {
+    fn auto_detect(raw_input_text: &str) -> NewlineStyle {
+        if let Some(pos) = raw_input_text.find('\n') {
+            let pos = pos.saturating_sub(1);
+            if let Some('\r') = raw_input_text.chars().nth(pos) {
+                NewlineStyle::Windows
+            } else {
+                NewlineStyle::Unix
+            }
+        } else {
+            NewlineStyle::Native
+        }
+    }
+
+    fn native() -> NewlineStyle {
+        if cfg!(windows) {
+            NewlineStyle::Windows
+        } else {
+            NewlineStyle::Unix
+        }
+    }
+
+    /// Apply this newline style to the formatted text. When the style is set
+    /// to `Auto`, the `raw_input_text` is used to detect the existing line
+    /// endings.
+    ///
+    /// If the style is set to `Auto` and `raw_input_text` contains no
+    /// newlines, the `Native` style will be used.
+    pub(crate) fn apply(self, formatted_text: &mut String, raw_input_text: &str) {
+        use NewlineStyle::*;
+        let mut style = self;
+        if style == Auto {
+            style = Self::auto_detect(raw_input_text);
+        }
+        if style == Native {
+            style = Self::native();
+        }
+        match style {
+            Windows => {
+                let mut transformed = String::with_capacity(formatted_text.capacity());
+                for c in formatted_text.chars() {
```

Comment:
I think you could you could use `replace` here rather than implementing yourself (we are guaranteed there are no `\r`s already in the formatted text).

Reply:
The newline conversion logic was moved from `formatting.rs`. I started to change it but I hit this debug assertion when running the unit tests:


```
test test::system_tests ... FAILED
...
thread 'test::system_tests' panicked at 'assertion failed: !formatted_text.contains('\r')', src\config\options.rs:151:9
```

```rust
debug_assert!(!formatted_text.contains('\r'));
match style {
    Windows => {
        *formatted_text = formatted_text.replace('\n', "\r\n");
    }
    Unix => return,
    Native => unreachable!("NewlineStyle::Native"),
    Auto => unreachable!("NewlineStyle::Auto"),
}
```